nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java

```diff
@@ -253,31 +226,16 @@ public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFacto
                     }
                 }
 
-                scriptEngine.setBindings(bindings, ScriptContext.ENGINE_SCOPE);
-
-                // Execute any engine-specific configuration before the script is evaluated
-                ScriptEngineConfigurator configurator =
-                        scriptingComponentHelper.scriptEngineConfiguratorMap.get(scriptingComponentHelper.getScriptEngineName().toLowerCase());
-
-                // Evaluate the script with the configurator (if it exists) or the engine
-                if (configurator != null) {
-                    configurator.init(scriptEngine, scriptToRun, scriptingComponentHelper.getModules());
-                    configurator.eval(scriptEngine, scriptToRun, scriptingComponentHelper.getModules());
-                } else {
-                    scriptEngine.eval(scriptToRun);
-                }
+                scriptRunner.run(bindings);
 
                 // Commit this session for the user. This plus the outermost catch statement mimics the behavior
                 // of AbstractProcessor. This class doesn't extend AbstractProcessor in order to share a base
                 // class with InvokeScriptedProcessor
                 session.commitAsync();
+                scriptingComponentHelper.scriptRunnerQ.offer(scriptRunner);
             } catch (ScriptException e) {
-                // Reset the configurator on error, this can indicate to the configurator to recompile the script on next init()
-                ScriptEngineConfigurator configurator =
-                        scriptingComponentHelper.scriptEngineConfiguratorMap.get(scriptingComponentHelper.getScriptEngineName().toLowerCase());
-                if (configurator != null) {
-                    configurator.reset();
-                }
+                // Create a new ScriptRunner to replace the one that caused an exception
+                scriptingComponentHelper.setupScriptRunners(1, scriptToRun, getLogger());
```

Comment:
If I understand correctly, if we have multiple ScriptRunners in the queue, but one causes an exception, we get rid of all the ScriptRunners in the queue and create a queue with only 1 ScriptRunner in it. Do we not need maxTasks number of them?

Reply:
The idea was to not offer the troublesome one back to the queue, but instead create a new one and offer that, so we'd always keep the queue full with runners that are hopefully good.

Reply:
That makes perfect sense. However, ExecuteScript has 1 instance of ScriptingComponentHelper. If the ExecuteScript processor is set to use 3 concurrent tasks, then 3 ScriptRunner instances get stored in the scriptingComponentHelper.scriptRunnerQ queue. If one ScriptRunner causes an exception, the scriptingComponentHelper.scriptRunnerQ queue gets replaced with a new queue with 1 element in it. Or am I missing something here?

Reply:
The successful tasks offer their runner back to the queue, the unsuccessful ones should replace the runner with a new one